src/test/java/com/uber/cadence/workflow/WorkflowTest.java

```diff
@@ -5194,6 +5194,59 @@ public void testSagaParallelCompensation() {
     assertTrue(tracer.getTrace().contains("executeActivity TestActivities::activity2"));
   }
 
+  public static class TestSignalExceptionWorkflowImpl implements TestWorkflowSignaled {
+    private boolean signaled = false;
+
+    @Override
+    public String execute() {
+      Workflow.await(() -> signaled);
+      return null;
+    }
+
+    @Override
+    public void signal1(String arg) {
+      for (int i = 0; i < 100; i++) {
+        Async.procedure(() -> System.out.println("test"));
+      }
+
+      throw new RuntimeException("exception in signal method");
+    }
+  }
+
+  @Test
+  public void testExceptionInSignal() throws InterruptedException {
+    startWorkerFor(TestSignalExceptionWorkflowImpl.class);
+    TestWorkflowSignaled signalWorkflow = workflowClient.newWorkflowStub(
+        TestWorkflowSignaled.class, newWorkflowOptionsBuilder(taskList).build());
+    CompletableFuture<String> result = WorkflowClient.execute(signalWorkflow::execute);
+    signalWorkflow.signal1("test");
+    try {
+      result.get(1, TimeUnit.SECONDS);
+    } catch (Exception e) {
+      // exception expected here.
+    }
+
+    // Suspend polling so that decision tasks are not retried. Otherwise it will affect our thread count.
+    if (useExternalService) {
+      workerFactory.suspendPolling();
+    } else {
+      testEnvironment.getWorkerFactory().suspendPolling();
+    }
+
+    // Wait for decision task retry to finish.
+    Thread.sleep(10000);
```

Comment:
Why do we retry decision task on exception? We should retry it only on Error.

Reply:
Because we want user to fix the code and then the execution can continue, right?

Reply:
Took a closer look, it seems that signal thread/handling is quite different than workflow thread... So here we are not catching the Exception and convert it to WorkflowExecutionException.

Reply:
So the behavior is different between signal method and workflow method, but this can be a separate issue though...

Reply:
OK. Would you file an issue to review the signal method error handling?